internal/runners/config/rules.go

```diff
@@ -0,0 +1,36 @@
+package config
+
+import "github.com/ActiveState/cli/internal/constants"
+
+type event func() error
+
+var emptyEvent = func() error { return nil }
+
+type configType int
+
+const (
+	String configType = iota
+	Int
+	Bool
+)
+
+type configRule struct {
+	allowedType configType
+	getEvent    event
+	setEvent    event
+}
+
+type configRules map[Key]configRule
+
+func (c configRules) Get(key Key) configRule {
+	rule, ok := c[key]
+	if !ok {
+		return configRule{String, emptyEvent, emptyEvent}
+	}
+	return rule
+}
+
+var rules = configRules{
+	constants.SvcConfigPid:  {Int, emptyEvent, emptyEvent},
+	constants.SvcConfigPort: {Int, emptyEvent, emptyEvent},
```

Comment:
This would require the events to be registered from inside this package. The idea is that whatever package "owns" the config key can register its usage from its own package.

Reply:
The svc one is kind of interesting because it is only used in the state-svc internal packages, which obviously wouldn't be able to register stuff that gets included in the `state` runtime. Maybe for those we could use the `svcmanager`.

This also raises an interesting point in that no one should ever be manually setting these config values. Maybe we need a way to say "set is not allowed for these keys". If that's cheaply added then I'd say go for it, but otherwise we can do a followup. Pretty unlikely someone would be running this in the interim.

Reply:
I started moving things around and ended up putting the `rules` in the `internal/config` package like @daved originally suggested. Not sure if this is what you intended from your comment, but take a look and let me know.

> This also raises an interesting point in that no one should ever be manually setting these config values. Maybe we need a way to say "set is not allowed for these keys". If that's cheaply added then I'd say go for it, but otherwise we can do a followup. Pretty unlikely someone would be running this in the interim.

I'll see what if I can come up with something that might work but if that ends up taking a lot of time I will file a follow up.

Reply:
Just quickly (sorry if it's not the clearest info) - We'd likely want to inject this from the main package. That is, configure the configuration. Alternately, if that information that affects the config config is not needed elsewhere, it can be set as internal/config package global. The disallowed keys as well as callbacks can be configured in this way.

Reply:
To be clear; I want to configure things like `constants.SvcConfigPort: {Int, emptyEvent, emptyEvent}` from the package that "owns" this aspect of the configuration. Eg. `SvcManager`.

Neither `main` nor `config` should have opinions on `SvcConfigPort` or other similar config keys.

Reply:
In this case, I suggest that the need is disparate from the config enough to warrant a different type. I thought it easiest to express this in a play: https://go.dev/play/p/sqFallRBDOR

This would mean that the underlying config is still passed around and whatnot, but logic/runners further down in the stack can wrap that config and have their own sets of rules. The keys can be codified in a central package and whatnot for ease.